Switch development to NixOS 22.05

dpausp opened the pull request:
@flyingcircusio/release-managers

### Significant breaking changes

- logrotate roles are now checked at build-time. Duplicate rules may lead to system build failures.
- remove global packages: php (when LAMP not enabled), nodejs, python2Full, usbutils, cyrus_sasl, cups, bundler, imagemagick, libjpeg, limsmbios, libtiff, libxml2, libxslt, openldap, protobuf


### Possible problems

* Some app deployments may rely on globally installed packages that have been removed. Deployments should manage their dependencies themselves, preferably by using `buildEnv` or adding packages globally using `environment.systemPackages` custom NixOS config in `/etc/local/nixos`.

### Significant package updates

- glibc 2.33 -> 2.34
- gnupg: 2.2.27 -> 2.3.6
- haproxy: 2.3.x -> 2.5.x
- nginx: 1.20.2 -> 1.22
- linux: 5.10.x -> 5.15.x
- logrotate: 3.18.x -> 3.20.x
- mailutils: 3.13 -> 3.14
- nix: 2.3 -> 2.8
- Added php 8.1
- openldap: 2.4.x -> 2.6.x
- Added openldap_2_4
- openssh: 8.8 -> 9.0
- podman: 3.4.x -> 4.1.x
- redis: 6.2.x -> 7.0.x
- solr: 8.6.x -> 8.11.x
- systemd: 249 -> 250
- telegraf: 1.20.x -> 1.22.x
- varnish: 7.0.x -> 7.1.x

## Other notable changes

- mongodb34 role has been removed. mongodb36 is EOL should not be used anymore.
- PHP 8.1 can now be used for the LAMP role.
- redis4 and rabbitmq38 role aliases were removed. Use redis and rabbitmq roles instead.
- The logrotate module also has been updated to freeform syntax: [services.logrotate.paths](https://nixos.org/manual/nixos/stable/options.html#opt-services.logrotate.paths) and [services.logrotate.extraConfig](https://nixos.org/manual/nixos/stable/options.html#opt-services.logrotate.extraConfig) will work, but issue deprecation warnings and [services.logrotate.settings](https://nixos.org/manual/nixos/stable/options.html#opt-services.logrotate.settings) should now be used instead. Platform logrotate config has been converted to the new style.

## Release process

Impact:

Changelog:

(announce in changelog that there's a new platform version and main development has switched to it, add release notes based on this PR message)

## Security implications

- [x] [Security requirements](https://wiki.flyingcircus.io/System_Development_Guideline#Security_requirement_principles_and_testing) defined? (WHERE)
  - we want to follow NixOS releases to get upstream security fixes
  - look at release notes to find new security implications and possible breakage when upgrading systems
- [x] Security requirements tested? (EVIDENCE)
  - reviewed NixOS release notes for breaking changes, security-related and other significant changes 
  - all tests are green, various VMs already running on 22.05 without problems

dpausp committed:
Trying to fix evaluation error caused by nix-phps

import generic.nix via relative path because importing from the Nix
store is not allowed in restricted eval mode.

dpausp committed:
Update nixpkgs to a nixos-unstable (pre 22.05) version

dpausp committed:
logrotate: disable config check

Complains about a duplicate, make it build and fix this later.

dpausp committed:
Update nixpkgs

osnyx committed:
percona80: 8.0.26-17 -> 8.0.27-18

- disable deprecated TokuDB storage engine

dpausp committed:
Update nixpkgs, fixes overriding github-runner version

osnyx committed:
update nix-phps to fix php73 build

notable package updates:
* php73: fix build
* composer: init at 2.2
* php74: 7.4.20 -> 7.4.29
* php80: 8.0.7 -> 8.0.18
* php81: init at 8.1.5

osnyx committed:
fix whitespace in nix-phps patches

it works both ways, but now the pre-commit checks are happy again

osnyx committed:
pre-commit: allow whitespace in vendored nix-phps

Revert "fix whitespace in nix-phps patches"

This reverts commit f7e04f9c8bcb67caf3e9b576f17f443e13f7ddad.

dpausp committed:
Update nixpkgs

osnyx committed:
workaround: disable all failing tests to get a first pre-release image

- this commit comments out all currently failing tests and packages
  to make Hydra generate a first release image
- these tests need to be fixed and re-enabled befor release!

PL-130551

dpausp committed:
Disable failing channels test

dpausp committed:
LAMP: add PHP 8.1

dpausp committed:
Update nixpkgs

We're at 22.05-beta now.

dpausp committed:
beats: fix journal and audit log shipping

* beats: 7.16.1 -> 7.17.4

Fixes auditbeat and filebeat crashing when trying to create a new
thread.

dpausp committed:
percona80: 8.0.27-18 -> 8.0.28-19, fix ldap error

Fixes a build problem with missing ldap_r after the last
nixpkgs update which included openldap 2.6. This brings back
openldap 2.4 from before the update and uses it for percona.

openldap_2_4.nix is a copy of pkgs/development/libraries/openldap/default.nix
from nixpkgs commit 3582135fe9589d0c823309cd38bebef37de369fd

dpausp committed:
xtrabackup: 8.0.26-18 -> 8.0.28-21

dpausp committed:
Update nixpkgs

dpausp committed:
Fix logrotate config for rsyslogd

Settings to reload rsyslogd after /var/log/wtmp and utmp rotation
conflicted with settings from nixpkgs. Our rules likely served no
purpose anymore, just remove them. The remaining settings for haproxy
are converted to the new settings option because extraConfig is
deprecated.

logging.nix test runs now.

dpausp committed:
Don't check solr for new, it's marked as insecure upstream

dpausp committed:
Fix deprecated acme mail option

dpausp committed:
Fix renamed ssh challengeResponseAuthentication option

dpausp committed:
logrotate: replace deprecated extraConfig text with settings attrset

dpausp committed:
Update nixpkgs

dpausp committed:
nginx: fix build

We add modsecurity-nginx by default which was broken since the
move to pcre2.
Update nixpkgs to get the modsecurity update which fixes this.

dpausp committed:
Enable working graylog, percona and mongodb tests

dpausp committed:
nginx: add mainline 1.23, fix test, add fc-logrotate

* Add fc-logrotate script to run logrotate with the platform config.
  Helpful for the nginx test and manual interaction on machines.
* We need a different nginx version to test binary updates but
  upstream only has 1.22.0 at the moment so let's update mainline
  Nginx to 1.23.
* The moreheaders extension doesn't work anymore with 1.23, remove it.
  We only need the version for the test for now so that's ok.

dpausp committed:
Update nixpkgs

dpausp committed:
mongodb: allow 3.6, remove 3.4 role and 3.2 test code

* 3.6 is marked as insecure upstream but we still need it for upgrades.
* 3.4 is end-of-life for a long time, remove it.

dpausp committed:
Refactor config for allowed unfree and insecure packages, remove OpenSSL 1.0.2

Unite the two places where unfree and insecure packages had to
be configured by using nixpkgs-config.nix as common import.
Config in ./default.nix affects users of our customized nixpkgs
, for example when using `nix-build -A <pkg>`.
On the other hand, `nixpkgs.config` defined in `nixos/platform/default.nix`
affects system rebuilds and NixOS tests.

They previously disagreed so behaviour was different.

OpenSSL 1.0.2 is very outdated and should not be used anywhere now.

dpausp committed:
Update nixpkgs

dpausp committed:
Removed patched zlib for Java

The patch is in nixpkgs zlib now.

dpausp committed:
k3s: remove from overlay

upstream has a newer minor version of 1.23, let's use it.

dpausp committed:
statshost: fix statshost-global test, add enable_http2

Prometheus tried to use HTTP2 which fails in the test with the
http.server faking the stats source. Prometheus 2.35 has a
new option to disable HTTP2 which helps to fix the test.
Also remove an environment variable from the test which had no
effect.

dpausp committed:
tests: remove subtest which build the system

It's too hard to maintain the code that is needed
to sync the situation in the VM with the system built
on the outside. It now tries to build a kernel but there's
no obvious reason why this happens now.

We now have a series of smoke test VMs that can check that
building systems with local config works properly.

dpausp committed:
Fix wkhtmltopdf test and modernize it

poppler HTML output has changed and the old assertion didn't match
anymore. The PDF itself is exactly the same as on 21.11 when ignoring
the CreationDate so comparing it to a reference file makes sense.

dpausp committed:
Fix gitlab test, test with Postgresql 14

dpausp committed:
Fix fc image building

Needed a newer lkl/linux version which is compatible with current
xfsprogs and more RAM for the image building VM.

Allow cptofs to use more RAM by patching the C code that starts
the kernel.

cptofs limits the memory usable by the kernel to 100M which
seems to be enough when I run the image build manually but not
when run by Hydra. Doesn't really make sense but allowing
cptofs to use 500M of the available RAM won't hurt anyway.

dpausp committed:
Update nixpkgs

dpausp committed:
Update release script and fix some shellcheck warnings

dpausp committed:
mailserver: fix reference to removed mime-types in roundcube config

dpausp committed:
redis: use default server for our role

Fixes the deprecation warning for now. We should look into
application-specific redis instances later.

dpausp committed:
set state version to 22.05

We cannot really use stateVersion like it is intended on NixOS, just
set it to the version of our platform to get rid of the warning.

dpausp commented on nixos/services/logrotate/default.nix:
```diff
@@ -74,6 +74,7 @@ in
       services.logrotate = {
         enable = true;
         extraConfig = mkOrder 50 globalOptions;
+        checkConfig = false;
```
do we need this or does it work now with build-time checks? If not, add a comment why we had to disable it.

dpausp replied:
checking works now with the default config, removed the setting

dpausp commented on pkgs/beats.nix:
```diff
@@ -0,0 +1,57 @@
+{ lib, fetchFromGitHub, buildGoModule, libpcap, nixosTests, systemd }:
```
can we upstream this?

dpausp replied:
Not really, upstream versions follow the ELK stack which is not relevant to us anymore as we fixed the ELK version to 7.10.2 to avoid running into licensing issues with Elasticsearch/Kibana. We need newer beats versions and updating the whole ELK stack upstream would be a lot of effort for us.

dpausp commented on tests/statshost/statshost-global.nix:
```diff
@@ -155,13 +155,10 @@ in {
         services.telegraf.enable = true;  # set in infra/fc but not in infra/testing
 
         flyingcircus.roles.statshost.prometheusLocationProxyExtraSettings = {
+          enable_http2 = false;
```
add a comment why this is needed here (python http.server)

osnyx commented on pkgs/overlay.nix:
```diff
@@ -192,7 +192,7 @@ in {
   });
 
-  filebeat7-oss = super.filebeat7.overrideAttrs(a: a // {
+  filebeat7-oss = self.filebeat7.overrideAttrs(a: a // {
```
Possibly squash with other beats commit.

osnyx commented on nixos/services/logrotate/default.nix:
```diff
@@ -3,10 +3,14 @@
 
-  globalOptions = ''
-    # Global default options for the Flying Circus platform.
+  # These header is only used for user-/application-specific logrotate config
```
typo: either *This header* or *These headers*

osnyx commented on tests/wkhtmltopdf.nix:
```diff
@@ -124,24 +124,28 @@ let
+    machine.copy_from_vm("${outputPdf}", "pdf")
+    machine.succeed("diff ${outputPdf} ${./wkhtmltopdf-expected.pdf}")
     machine.succeed('${pkgs.poppler_utils}/bin/pdftohtml -s -fontfullname /tmp/sample1.pdf')
     _, output = machine.execute('cat sample1-html.html')
```
Do we still need that `output`, does printing it provide valuable insights? Because the assert depending on it later has been removed.

osnyx reviewed (approved):
Overall the changes of this PR look good.

A bit of cleanup like squashing related commits and their later fixups, e.g. the image building commits, as well as some explanatory comments for certain changes can make this PR even more tidy.

dpausp committed:
Remove unused vendored gitlab package

dpausp committed:
Remove unused vendored grub package

dpausp committed:
Remove bash_5_1_p12 package, upstream bash is newer

dpausp committed:
Remove kubernetes-helm override, upstream is newer

dpausp committed:
Remove obsolete collectd fix

dpausp committed:
solr: remove override to include patched zlib

zlib fix is in nixpkgs now, no need to override Java anymore.

dpausp committed:
Remove flannel from overlay, upstream is newer

dpausp committed:
Remove cgmemtime package

cgmemtime was a nice tool to find out how much memory is consumed
by a process and its childen. It doesn't work anymore with cgroups v2
anymore so it's useless now.

dpausp committed:
Remove unused zlib patch

dpausp committed:
hydra config: update platform version to 22.05

dpausp committed:
nixpkgs-config: change comment about SSPL/mongodb

We could allow SSPL generally on 22.05 but we only use mongodb at the
moment. Keeping it as it is avoids accidentally adding dependencies on
SSPL-licensed software. The implications of using SSPL code is still not
clear so it's better to avoid it by default.

dpausp committed:
postgresql: remove obsolete fix for postgis problem, cleanup

Postgis seems to work properly in the test now for 12, so we can
simplify the test script.

Replace rolename with a version string. This is possible now because
the last version with the old naming scheme, 9.6, got dropped and
versions start at 10 now. We use a version argument most of the time,
it makes version comparisons easier and is less to type.

dpausp committed:
Replace non-code occurences of 21.11 with 22.05

dpausp committed:
Update nixpkgs

dpausp committed:
Remove role aliases for redis4 and rabbitmq38

We had them for a long time and they produce annoying deprecation
warnings when building the system, coming from the devhost role.

dpausp committed:
Remove globally installed PHP when LAMP is not enabled

The LAMP role, when enabled, still installs the matching PHP
version globally.

dpausp committed:
Remove globally installed nodejs

It's big and installing it by default in systemPackages makes
it harder to get another nodejs version globally.

dpausp committed:
Remove some global pkgs or replace with non-X variants

This reduces system closure size by about 500MB for basic VMs.

Adding packages to environment.systemPackages is easy today with
/etc/local/nixos config and many of the tools can be also ran via
nix-shell if needed. App dependencies should be part of app deployments
and not be expected to be present globally.

* python2Full is big (pulls in a lot of graphics stuff) and should not
  be used anymore.
* Installing libraries globally has no effect anymore.
* links2 and latencytop now don't depend on graphics libraries anymore.

Some X libraries are still in the base system because they cannot
be easily removed from everything. There's an `environment.noXlibs`
option but it has too many side-effects and needs a lot of custom
compilation on our Hydra.

dpausp committed:
shell.nix: fix NIX_PATH in shell, use sudo in nixos-repl

Just passing a env var to mkDerivation is not enough to change the
NIX_PATH inside the shell when run by nix-shell. We have to
export it from the shellHook.

Before, this worked only for nix-direnv.

We have to run nixos-repl.nix as root, adding sudo to the nixos-repl
command is more convenient. Using `-E` keeps the environment to
use the correct NIX_PATH set by the shell.

dpausp committed:
jitsi: remove version overrides, upstream has the same versions

We still keep the openssl 3 fix. Upstream it later.

dpausp committed:
LAMP: remove prce2 overrides

Upstream version of prce2 is now 10.40 and our regression test
is green. Before, we had problem with 10.37 and had to downgrade
to 10.35.

dpausp committed:
Remove unused wkhtmltopdf 0.12.4 package

dpausp committed:
wkhtmltopdf: fix configureFlags deprecation warning
